### PR TITLE
tools/ceph-objectstore-tool: tricks to tolerate disk errors for "pg export" command.

### PR DESCRIPTION
**Pull request by ifed01**
Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by ifed01** [comment]
jenkins test make check

**Comment by ifed01** [comment]
jenkins test doc

**Comment by ifed01** [comment]
jenkins test render docs

**Comment by ifed01** [comment]
jenkins render docs

**Comment by ifed01** [comment]
jenkins test docs

**Comment by pereman2** [comment]
jenkins test make check

**Comment by aclamk** [comment]
jenkins test make check

**Comment by ifed01** [comment]
jenkins test make check

File: src/tools/ceph_objectstore_tool.cc [hunk 1]
```diff
@@ -3438,6 +3441,9 @@ int main(int argc, char **argv)
       "Threshold (in seconds) to consider omap listing slow (for op=list-slow-omap)")
     ("dump-data-dir", po::value<string>(&dump_data_dir),
      "Directory to dump object data (for op=dump-export)")
+    ("log-file,l", po::value<string>(&log_file), "log file")
+    ("log-level", po::value<int>(&log_level), "log level (30=most, 20=lots, 10=some, 1=little)")
+    ("conf,c", po::value<string>(&conf_file), "config file")
```
Comment (aclamk) [inline src/tools/ceph_objectstore_tool.cc:3446]: I was able to pass "--conf=my-ceph.conf" to regular COT. This is redundant.

File: src/tools/ceph_objectstore_tool.cc [hunk 2]
```diff
@@ -3438,6 +3441,9 @@ int main(int argc, char **argv)
       "Threshold (in seconds) to consider omap listing slow (for op=list-slow-omap)")
     ("dump-data-dir", po::value<string>(&dump_data_dir),
      "Directory to dump object data (for op=dump-export)")
+    ("log-file,l", po::value<string>(&log_file), "log file")
```
Comment (aclamk) [inline src/tools/ceph_objectstore_tool.cc:3444]: I can pass "--log-file=zz.txt" to regular COT. This is redundant.

File: src/os/bluestore/bluestore_tool.cc [hunk 3]
```diff
@@ -354,6 +358,27 @@ int main(int argc, char **argv)
     std::cerr << e.what() << std::endl;
     exit(EXIT_FAILURE);
   }
+  if (!action_aux.empty()) {
```
Comment (aclamk) [inline src/os/bluestore/bluestore_tool.cc:361]: I have a proposal to collapse ifs:
```
if (action_aux != action) {
  ... cerr ... 
  EXIT;
}
if (action.empty()) {
  action.swap(action_aux);
}
```

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -8827,6 +8827,131 @@ bool BlueStore::has_null_manager() const
...
+  });
+
+  _kv_start();
+  auto stop_kv = make_scope_guard([&] {
+    if (!mounted) {
+      _kv_stop();
+    }
+  });
+
+  r = _deferred_replay();
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:8853]: I have a serious doubt if `_deferred_replay()` is compatible with read-only.
In kv-sync-thread we have:
```
for (auto b : deferred_stable) {
....
	  synct->rm_single_key(PREFIX_DEFERRED, key);
....	  
      }
```
My claim is that if there were any deferreds, we will fail.
So maybe it is better not to do it at all?

File: src/os/bluestore/BlueStore.cc [hunk 5]
```diff
@@ -8827,6 +8827,131 @@ bool BlueStore::has_null_manager() const
...
+  if (r < 0) {
+    return r;
+  }
+  mempool_thread.init();
+  mounted = true;
+  return r;
+}
+
+int BlueStore::_mount_readonly()
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:8862]: I feel a need of some comment of how this read-only is achieved.
I guess it is some reliance on the fact that if allocator is not initialized then nothing new can be written?

**Review by aclamk — COMMENTED** [review]
I think we need to address problem of deferred_replay().
Some unit test / scenario script would do great to convince me that this is good as it is.

File: src/os/bluestore/BlueStore.cc (hunk 4 above)
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:8853]: submit_transaction_sync in kv_sync_thread is protected with db_was_opened_read_only flag now: 

int r = db_was_opened_read_only || cct->_conf->bluestore_debug_omit_kv_commit ?
	0 : db->submit_transaction_sync(synct);

File: src/os/bluestore/BlueStore.cc (hunk 5 above)
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:8862]: Not really - major actions for opening in RO mode are the follows:
- _open_db calls RocksDB::open_read_only if read_only param provided. Which in turn results in calling 
status = rocksdb::DB::OpenForReadOnly(opt, path, &db);
- omitting calling submit_transaction_sync in kv_sync_thread if opened in RO
- omitting sync_metadata call in BlueFS::umount if opened in RO mode.
- omitting allocator creation which is rather an auxiliary mean to ensure no writes to new target locations occur. Writing to already allocated extents, e.g. ones bound to deferred ops are allowed though.

File: src/tools/ceph_objectstore_tool.cc (hunk 1 above)
Comment (ifed01) [reply src/tools/ceph_objectstore_tool.cc]: Agree.

File: src/os/bluestore/bluestore_tool.cc (hunk 3 above)
Comment (ifed01) [reply src/os/bluestore/bluestore_tool.cc]: Done

**Comment by ifed01** [comment]
> I think we need to address problem of deferred_replay(). Some unit test / scenario script would do great to convince me that this is good as it is.

Test case provided in store_test.cc

**Review by aclamk — APPROVED** [review]
Cute test for deferred replay in read-only.

**Comment by NitzanMordhai** [comment]
@yuriweinstein rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomcSFxA3U2X1976-wip-yuri3-testing-2024-03-25-0744-old-wip-yuri3-testing-2024-03-24-1519-old-wip-yuri3-testing-2024-03-12-0851

**Pull request merged**